services/src/northbound/src/main/java/org/openkilda/northbound/messaging/kafka/KafkaMessagingChannel.java

```diff
@@ -0,0 +1,238 @@
+
+import org.openkilda.messaging.Message;
+import org.openkilda.messaging.error.ErrorMessage;
+import org.openkilda.messaging.error.MessageException;
+import org.openkilda.messaging.info.ChunkedInfoMessage;
+import org.openkilda.messaging.info.InfoData;
+import org.openkilda.messaging.info.InfoMessage;
+import org.openkilda.northbound.messaging.MessageProducer;
+import org.openkilda.northbound.messaging.MessagingChannel;
+import org.openkilda.northbound.messaging.exception.MessageNotSentException;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.collections4.map.PassiveExpiringMap;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.kafka.support.SendResult;
+import org.springframework.stereotype.Component;
+import org.springframework.util.concurrent.ListenableFuture;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import javax.annotation.PostConstruct;
+
+/**
+ * Implementation of {@link MessagingChannel} for kafka.
+ */
+@Component
+public class KafkaMessagingChannel implements MessagingChannel {
+
+    private static final Logger logger = LoggerFactory.getLogger(KafkaMessagingChannel.class);
+
+    @Value("${northbound.messages.expiration.minutes}")
+    private int expiredTime;
+
+    /**
+     * Requests that are in progress of processing.
+     */
+    private Map<String, CompletableFuture<InfoData>> pendingRequests = new ConcurrentHashMap<>();
+    private Map<String, CompletableFuture<List<InfoData>>> pendingChunkedRequests = new ConcurrentHashMap<>();
+
+    /**
+     * Chains of chunked messages, it is filling by messages one by one as soon as the next linked message is received.
+     */
+    private Map<String, List<ChunkedInfoMessage>> messagesChains;
+
+    @Autowired
+    private MessageProducer messageProducer;
+
+    /**
+     * Creates storage for chains of messages.
+     */
+    @PostConstruct
+    public void setUp() {
+        messagesChains = new PassiveExpiringMap<>(expiredTime, TimeUnit.MINUTES, new HashMap<>());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<InfoData> sendAndGet(String topic, Message message) {
+        CompletableFuture<InfoData> future = new CompletableFuture<>();
+
+        ListenableFuture<SendResult<String, Message>> futureResult = messageProducer.send(topic, message);
+        futureResult.addCallback(
+                sentResult -> pendingRequests.put(message.getCorrelationId(), future),
```

Comment:
From race condition point of view - better to do `pendingRequests.put(message.getCorrelationId(), future)` unconditionally i.e. even before `messageProducer.send(...)` and remove this record in case of error. In our case we already have clean via future object.